fix: dedup cilium endpoint controllers

sypakine opened the pull request:
Fixes: #30820

This prevents an Endpoint controller from deleting a CEP object that has been claimed by another Endpoint controller.

This can happen during endpoint creation if the container sandbox creation fails and is retried (using a new container ID). Each sandbox creation attempt starts a new Endpoint controller for the same namespaced/named CEP object.

If the first Endpoint controller has not deleted the CEP object before the second Endpoint controller fetches and patches the CEP object, the UID will remain the same for the object but with a different container ID. Therefore, the CEP object should be managed by the Endpoint controller associated with the current container. At this point, the original Endpoint controller should no longer delete the CEP object.

```release-note
Map endpoint from container recreation to existing cilium endpoint and endpoint controller.
```

christarazi reviewed (commented):
I think the fix seems reasonable here, but just to have a fully clear mental model of the issue, what does "synchronizer" and "another synchronizer" specifically refer to in your description? In the linked issue, you used the terms "Cilium endpoint manager `X`" and "Cilium endpoint manager `Y`", which I assume are stand-ins for "synchronizer" and "another synchronizer", respectively.

christarazi commented on pkg/endpointmanager/endpointsynchronizer.go:
```diff
@@ -461,8 +461,26 @@ func deleteCEP(ctx context.Context, scopedLog *logrus.Entry, ciliumClient v2.Cil
+	if e.GetContainerID() != fetchedEndpoint.Status.ExternalIdentifiers.ContainerID {
+		scopedLog.Info("Skipping CiliumEndpoint deletion because it has a different Container ID.")
+		return nil
+	}
```
Instead of doing this, have you considered calling `cepOwner.GetHostIP()` and checking against the local node's IP to determine whether the CEP is indeed owned by this node? If the node's IP matches that means that this "endpoint manager" (aka node) owns this CEP and can issue the delete.

sypakine replied:
Using a distilled version of https://gist.github.com/joestringer/086d3cff805d1e4dade93319bd1376d4 to help visualize this sequencing:

![image](https://github.com/cilium/cilium/assets/6053986/32191b71-91bf-43d5-99bb-72e944070493)

This issue happens within the same node, so the host IP remains constant.

---

So the problem is one or both of:

1. async Endpoint controllers for the same pod
  - if there was only a single controller, it could resolve this conflict

2. Endpoint controllers share k8s CEP object
  - if CEP object was not keyed using the namespaced-name of the Pod, each controller would create and manage the life-cycle of its own CEP

Both of these approaches would need design. The question is whether the fix for this issue should be incumbent on larger design changes.

christarazi replied:
Thanks for the helpful diagram. I agree with you on the problems listed above and I agree that problem (2) seems likely that it requires a top-down, larger design change, so let's ignore that one for now. I believe we might have some wiggle room if we explore (1) a bit more because I think there's a potential approach that I see.

So it seems that the root issue is that we are launching multiple controllers to essentially manage the same CEP. This lifecycle assoication seems wrong if we are to take the text of the endpointsynchronizer godoc:

> // RunK8sCiliumEndpointSync starts a controller that synchronizes the endpoint
// to the corresponding k8s CiliumEndpoint CRD. **It is expected that each CEP**
// **has 1 controller that updates it**, and a local copy is retained and only
// updates are pushed up.
// CiliumEndpoint objects have the same name as the pod they represent.

The key part is 

> It is expected that each CEP has 1 controller that updates it

because we are not holding up this assumption in the code. Right now, the code is tying the lifecycle of the controller to the internal endpoint object, rather than the CEP.

My thinking is 

1. Change the endpointsynchronizer object (`EndpointSynchronizer`) to become a central/global object that itself to manages these controllers, rather than each individual endpoint object (`Endpoint`).
2. Change the controller names from using `ep.ID` (endpoint IDs) to the namespace/name of the CEP, so that we can create a 1:1 relationship between controllers and CEPs, rather than 1:1 relationship with controllers `Endpoint`s.

The above changes should allow us to hold up the assumption aforementioned assumption.

For (1), the individual endpoint objects can have a reference to the endpointsynchronizer global object and request to create/delete controllers. When it makes these requests, the individual endpoints can pass their ID so that the endpointsynchronizer can disambiguate between a create and delete on the same CEP. Because it is able to disambiguate which "owner" (endpoint object) is requesting a create or delete, that allow us to detect the case presented in the diagram where essentially the creation of 2nd endpoint object (that triggers the creation of "SynchronizerY") **should cancel** the deletion request of "SynchronizerX". What I'm thinking there is probably the usage of `context.Context` to actually perform that cancellation. 

I recognize this might be a bit in the weeds so I'm happy to clarify further. I feel like it's worth exploring this path, because ultimately I feel a bit uneasy about accepting an additional call to apiserver on deletion. It's especially expensive in clusters that have a lot of pod churn.

sypakine replied:
Thanks @christarazi. Perhaps I could answer this question for myself with more context on the `EndpointSychronizer` implementation, but if we key internal `Endpoint`s on the Pod namespace/name, mapping the `Endpoint` with the `CiliumEndpoint` object (2. from above), would we need to change the synchronizer (1.) that significantly?

christarazi replied:
I'm not sure I fully understood your question, but as far as changing the endpointsynchronizer, the logic itself does not need to change. Rather how the object is instantiated and its lifecycle is what needs to change, i.e. from multiple instances to a central instance.

sypakine committed:
fix: dedup endpoint controller

- map endpoint controller to CEP name, not ID
- reuse endpoint ID for endpoint creation

During endpoint creation, CRI failure may result in multiple containers for a single Pod/CEP object. Subsequent CNI ADD calls for the Pod will contain a new creation parameters (primarily IP and Container ID).

This change dedupes these CNI add events, mapping to the existing endpoint controller and CEP object.

Signed-off-by: Mark St John <markstjohn@google.com>

christarazi commented on pkg/endpointmanager/manager.go:
```diff
@@ -653,6 +653,13 @@ func (mgr *endpointManager) AddEndpoint(owner regeneration.Owner, ep *endpoint.E
 	}
 
+	oldEP := mgr.LookupCEPName(ep.GetK8sNamespaceAndPodName())
+	if oldEP != nil {
+		// This is a container recreation event for an existing Pod.
+		// We want to merge this new endpoint with the existing endpoint.
+		ep.ID = oldEP.ID
+	}
```
Is that all that's really needed to "merge" the new endpoint with the existing?

sypakine replied:
Based on my reproduction scenario, yes this replaces the endpoint controller with the new endpoint associated with the CNI ADD. The controller name change along was not sufficient.

christarazi replied:
My suspicion is only setting the ID probably leads to memory leaks due to the previous Endpoint references not being cleaned up/replaced everywhere. Have you looked into that?

aanm replied:
I agree with Chris on this. We are basically re-using an existing endpoint ID without waiting for all the instances of the "old" endpoint to be removed. I think this might cause some unknown issues. Is this change really necessary? The `EndpointSyncControllerName(name string)` would make the controller "unique" for the pod name, isn't that enough?

sypakine replied:
That was my first approach, but a second controller was still being spawned for the second create. This change deduped the controller, but is running afoul of the docker pod creation during the integration tests.

Let me try removing this again and track down an alternative that doesn't reuse the endpoint ID.

aanm reviewed (changes requested):
@sypakine thank you for the patch. Sorry for taking so much time to leave a review.
